dask/order.py

```diff
@@ -274,177 +276,460 @@ def finish_now_key(x):
+                            if len(finish_now) > 1:
+                                finish_now = sorted(finish_now, key=finish_now_key)
+                            for dep in finish_now:
+                                result[dep] = i
+                                i += 1
+                            add_to_inner_stack = False
+                        elif skip_root_node:
+                            for dep in root_nodes:
+                                num_needed[dep] -= 1
+                                # Use remove here to complain loudly if our assumptions change
+                                deps.remove(dep)  # Safe to mutate
+
+                    if deps:
+                        for dep in deps:
+                            num_needed[dep] -= 1
+                        process_singles = False
                     else:
-                        inner_stack.extend(deps)
-                    seen_update(deps)
+                        continue
+            elif inner_stacks:
+                inner_stack = inner_stacks_pop()
+                inner_stack_pop = inner_stack.pop
+                continue
+            elif singles:
+                process_singles = True
+            elif later_singles:
+                # No need to be optimistic: all nodes in `later_singles` will free a dependency
+                # when run, so no need to check whether dependents are in `seen`.
+                deps = set()
+                for single in later_singles:
+                    if single in result:
+                        continue
+                    while True:
+                        dep2 = dependents[single]
+                        result[single] = i
+                        i += 1
+                        if metrics[single][3] == 1:  # min_height
+                            # Don't leave any dangling single nodes!  Finish all dependents that are
+                            # ready and are also root nodes.
+                            finish_now = {
+                                dep
+                                for dep in dep2
+                                if not dependents[dep] and num_needed[dep] == 1
+                            }
+                            if finish_now:
+                                dep2 -= finish_now  # Safe to mutate
+                                if len(finish_now) > 1:
+                                    finish_now = sorted(finish_now, key=finish_now_key)
+                                for dep in finish_now:
+                                    result[dep] = i
+                                    i += 1
+                            elif skip_root_node:
+                                for dep in root_nodes:
+                                    num_needed[dep] -= 1
+                                    # Use remove here to complain loudly if our assumptions change
+                                    dep2.remove(dep)  # Safe to mutate
```

Comment:
Not covered.  Why not?  Are there not that many examples where `skip_root_node` is True, or is this somehow guaranteed not to run?